Fix the warnings for extern-template in JSObject template classes

iidmsa opened the pull request:
#### 93d1c5aab08094e0b241f15389a7c00e341c5604
<pre>
Fix the warnings for extern-template in JSObject template classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=161979">https://bugs.webkit.org/show_bug.cgi?id=161979</a>
<a href="https://rdar.apple.com/problem/158907443">rdar://problem/158907443</a>

Reviewed by Keith Miller.

* Source/JavaScriptCore/API/JSAPIWrapperObject.mm:
* Source/JavaScriptCore/API/JSCallbackObject.cpp:
(JSC::JSCallbackObject&lt;JSGlobalObject&gt;::visitChildren):
(JSC::JSCallbackObject&lt;JSNonFinalObject&gt;::visitChildren):
(JSC::JSCallbackObject&lt;JSAPIWrapperObject&gt;::visitChildren):
* Source/JavaScriptCore/API/JSCallbackObject.h:
* Source/JavaScriptCore/API/JSCallbackObjectFunctions.h:
(JSC::JSCallbackObject&lt;Parent&gt;::finishCreation):

Canonical link: <a href="https://commits.webkit.org/299144@main">https://commits.webkit.org/299144@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5d4ec9f8b445642624f8d5b00073418ab8616f4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118009 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37687 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28323 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124158 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70044 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38380 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46269 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/89551 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120961 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30567 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105819 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70043 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29634 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23936 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67821 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110120 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99990 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24114 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127240 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/116519 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44912 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33841 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/98223 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45273 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102043 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98009 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24930 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43409 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21398 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44784 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50458 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145217 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44244 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/37358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47589 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45933 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/API/JSCallbackObjectFunctions.h:
```diff
@@ -104,7 +104,10 @@ void JSCallbackObject<Parent>::finishCreation(VM& vm)
+    if constexpr (std::is_same_v<Parent, JSGlobalObject>)
+        init(jsCast<JSGlobalObject*>(this));
+    else
+        init(static_cast<Parent*>(this)->globalObject());
```
Why do you have this change? Presumably the parent inherits from JSGlobalObject since `ASSERT(Parent::isGlobalObject());`.

iidmsa replied:
Reverted.

kmiller68 commented on Source/JavaScriptCore/API/JSCallbackObject.h:
```diff
@@ -249,6 +249,69 @@ void JSCallbackObject<Parent>::visitChildrenImpl(JSCell* cell, Visitor& visitor)
 
+template<> const ClassInfo JSCallbackObject<JSGlobalObject>::s_info;
+template<> JSCallbackObject<JSGlobalObject>* JSCallbackObject<JSGlobalObject>::create(VM&, JSClassRef, Structure*);
+template<> Structure* JSCallbackObject<JSGlobalObject>::createStructure(VM&, JSGlobalObject*, JSValue proto);
```
I think you just need: 

```
template<> const ClassInfo JSCallbackObject<JSGlobalObject>::s_info;
template<> const ClassInfo JSCallbackObject<JSNonFinalObject>::s_info;

#if JSC_OBJC_API_ENABLED
template<> const ClassInfo JSCallbackObject<JSAPIWrapperObject>::s_info;
#endif
```

kmiller68 commented on Source/JavaScriptCore/API/JSAPIWrapperObject.mm:
```diff
@@ -73,8 +73,6 @@
 static JSC_DECLARE_CUSTOM_GETTER(callbackGetterJSAPIWrapperObjectCallbackObject);
 static JSC_DECLARE_CUSTOM_GETTER(staticFunctionGetterJSAPIWrapperObjectCallbackObject);
 
-DEFINE_VISIT_CHILDREN_WITH_MODIFIER(template<>, JSCallbackObject<JSAPIWrapperObject>);
```
I don't think you need this change.

iidmsa replied:
Reverted.

kmiller68 commented on Source/JavaScriptCore/API/JSCallbackObject.cpp:
```diff
@@ -41,9 +41,6 @@ static JSC_DECLARE_CUSTOM_GETTER(staticFunctionGetterJSNonFinalObjectCallbackObj
 static JSC_DECLARE_CUSTOM_GETTER(callbackGetterJSGlobalObjectCallbackObject);
 static JSC_DECLARE_CUSTOM_GETTER(staticFunctionGetterJSGlobalObjectCallbackObject);
 
-DEFINE_VISIT_CHILDREN_WITH_MODIFIER(template<>, JSCallbackObject<JSNonFinalObject>);
```
Ditto.

iidmsa replied:
Reverted.

kmiller68 commented on Source/JavaScriptCore/API/JSCallbackObject.cpp:
```diff
@@ -145,6 +142,62 @@ GCClient::IsoSubspace* JSCallbackObject<JSGlobalObject>::subspaceForImpl(VM& vm,
+    ASSERT_GC_OBJECT_INHERITS((static_cast<JSGlobalObject*>(thisObject)), JSCallbackObject<JSGlobalObject>::info());
+    JSGlobalObject::visitChildren(thisObject, visitor);
+    thisObject->m_callbackObjectData->visitChildren(visitor);
+}
```
Don't need these either.

iidmsa replied:
Reverted.

kmiller68 commented on Source/JavaScriptCore/API/JSCallbackObject.h:
```diff
@@ -249,6 +249,13 @@ void JSCallbackObject<Parent>::visitChildrenImpl(JSCell* cell, Visitor& visitor)
     thisObject->m_callbackObjectData->visitChildren(visitor);
 }
 
+template<> const ClassInfo JSCallbackObject<JSGlobalObject>::s_info;
```
Nit: Might be worth a comment explaining that `JSCallbackObject<T>::info()`'s forward definition triggers an undefined template var warning without these declarations in the same translation unit.

iidmsa replied:
Comment Added.

iidmsa committed:
Fix the warnings for extern-template in JSObject template classes
https://bugs.webkit.org/show_bug.cgi?id=161979
rdar://problem/158907443

Reviewed by Keith Miller.

* Source/JavaScriptCore/API/JSAPIWrapperObject.mm:
* Source/JavaScriptCore/API/JSCallbackObject.cpp:
(JSC::JSCallbackObject<JSGlobalObject>::visitChildren):
(JSC::JSCallbackObject<JSNonFinalObject>::visitChildren):
(JSC::JSCallbackObject<JSAPIWrapperObject>::visitChildren):
* Source/JavaScriptCore/API/JSCallbackObject.h:
* Source/JavaScriptCore/API/JSCallbackObjectFunctions.h:
(JSC::JSCallbackObject<Parent>::finishCreation):

Canonical link: https://commits.webkit.org/299144@main
